#### COMPANION WORKBOOK

# Pandas

To make the most out this program, we strongly recommend you to:
1. First practice writing and implementing all of the code from Coding Section of the online lesson.
2. Then, freely experiment with and explore any interesting or confusing concepts. Simply insert new code cells and then use the help of Google and official documentation.
3. Finally, tackle all of the exercises at the end. They will help you tie everything together and **learn in context.**

#### <span style="color:#555">LESSON CODE SANDBOX</span>

Use this space to practice writing and implementing all of the code from Coding Section of the online lesson. Insert new code cells as needed, and feel free to write notes to yourself in Markdown.

## I. Pandas DataFrames are like spreadsheets.

In [2]:
import pandas as pd


In [3]:
example_dataframe = pd.DataFrame({
    'column_1': [5, 4, 3],
    'column_2': ['a', 'b', 'c']
})

example_dataframe

column_1 column_2
0         5        a
1         4        b
2         3        c

In [4]:
# read dataset from CSV file

iris_df = pd.read_csv('project_files/iris.csv')

print(type(iris_df))

<class 'pandas.core.frame.DataFrame'>


In [5]:
# display first five rows of a dataset

iris_df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [6]:
# shape of iris_df

print(iris_df.shape)

(150, 5)


In [8]:
# Get numbers of rows in iris_df

print(len(iris_df))

print(iris_df.shape[0])

150
150


In [10]:
# Display Summary Statistics

print(iris_df.describe())

       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000


## II. Pandas Series are like single columns.

In [12]:
# create pandas series

int_series = pd.Series([0,1,2,3,4])

print(int_series)

print(type(int_series))

0    0
1    1
2    2
3    3
4    4
dtype: int64
<class 'pandas.core.series.Series'>


In [14]:
# selecting column / series from data frame

iris_series = iris_df.petal_length
print(iris_series)
print(type(iris_series))


iris_serie = iris_df['petal_length']
print(iris_serie)
print(type(iris_serie))

# check if above approaches are the same

print(all(iris_series == iris_serie))


0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: petal_length, Length: 150, dtype: float64
<class 'pandas.core.series.Series'>
0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: petal_length, Length: 150, dtype: float64
<class 'pandas.core.series.Series'>
True


In [19]:
# print first 5 items in iris_series

print(iris_series.head())


# minimum sepal length

print('minimum sepal length is {}'.format(iris_series.min()))


#maximum sepal length

print('maximum sepal length is {}'.format(iris_series.max()))




0    1.4
1    1.4
2    1.3
3    1.5
4    1.4
Name: petal_length, dtype: float64
minimum sepal length is 1.0
maximum sepal length is 6.9


In [20]:
# print unique species using unique() function

print(iris_df.species.unique())

['setosa' 'versicolor' 'virginica']


In [22]:
# series Math

print(iris_df.petal_length.median())

# using numpy
import numpy as np

print(np.median(iris_df.petal_length))

4.35
4.35


In [25]:
#feature engineering

iris_df['petal_area'] = iris_df.petal_length * iris_df.petal_width


print(iris_df.petal_area.head())

0    0.28
1    0.28
2    0.26
3    0.30
4    0.28
dtype: float64


## III. Boolean masks allow you to filter data.

In [26]:
# create pandas series list of boolean
list_mask = [True, False, False, True]

series_mask = pd.Series([True, False, False, True])

print(series_mask)

0     True
1    False
2    False
3     True
dtype: bool


In [27]:
# in real life a boolean mask is created by applying a conditional  statement to another pandas series 


example = pd.Series([10, 5, -3, 2])

print(example)


0    10
1     5
2    -3
3     2
dtype: int64


In [28]:
# create a boolean mask for all positive values in example using a condition

series_mask = example > 0

print(series_mask)

0     True
1     True
2    False
3     True
dtype: bool


In [29]:
# use the boolean mask to filter example and keep positive observations

print(example[series_mask])



0    10
1     5
3     2
dtype: int64


In [30]:
# inverted filtering


print(example[-series_mask])

2   -3
dtype: int64


In [32]:
# filtering rows from dataframe

# create mask

mask = iris_df.petal_area > 14

# index the mask in the data frame
print(iris_df[mask])

iris_df[mask]

     sepal_length  sepal_width  petal_length  petal_width    species  \
100           6.3          3.3           6.0          2.5  virginica   
109           7.2          3.6           6.1          2.5  virginica   
117           7.7          3.8           6.7          2.2  virginica   
118           7.7          2.6           6.9          2.3  virginica   
135           7.7          3.0           6.1          2.3  virginica   
144           6.7          3.3           5.7          2.5  virginica   

     petal_area  
100       15.00  
109       15.25  
117       14.74  
118       15.87  
135       14.03  
144       14.25  


sepal_length  sepal_width  petal_length  petal_width    species  \
100           6.3          3.3           6.0          2.5  virginica   
109           7.2          3.6           6.1          2.5  virginica   
117           7.7          3.8           6.7          2.2  virginica   
118           7.7          2.6           6.9          2.3  virginica   
135           7.7          3.0           6.1          2.3  virginica   
144           6.7          3.3           5.7          2.5  virginica   

     petal_area  
100       15.00  
109       15.25  
117       14.74  
118       15.87  
135       14.03  
144       14.25

In [35]:
# using boolean mask to create indicator variables

example_mask = pd.Series([ True, False, False, True, True])

example_mask.astype(int)

0    1
1    0
2    0
3    1
4    1
dtype: int64

In [38]:
# create indicator variable for iris_df.petal_area > 14

iris_df['giant'] = (iris_df.petal_area > 14).astype(int)

# get first 5 element of two series from iris_df
iris_df[['petal_area', 'giant']].head()

petal_area  giant
0        0.28      0
1        0.28      0
2        0.26      0
3        0.30      0
4        0.28      0

In [39]:
# Multiple masks using the & operator or set the multiple masks to different variable

# task 1: check if it is versicolor or verginica
specie_mask = iris_df.species.isin(['versicolor', 'virginica'])

# task 2: big width sepal
big_width_mask = iris_df.sepal_width > 3.2


# task 3: Combination of both masks
iris_df[specie_mask & big_width_mask]










sepal_length  sepal_width  petal_length  petal_width     species  \
56            6.3          3.3           4.7          1.6  versicolor   
85            6.0          3.4           4.5          1.6  versicolor   
100           6.3          3.3           6.0          2.5   virginica   
109           7.2          3.6           6.1          2.5   virginica   
117           7.7          3.8           6.7          2.2   virginica   
124           6.7          3.3           5.7          2.1   virginica   
131           7.9          3.8           6.4          2.0   virginica   
136           6.3          3.4           5.6          2.4   virginica   
144           6.7          3.3           5.7          2.5   virginica   
148           6.2          3.4           5.4          2.3   virginica   

     petal_area  giant  
56         7.52      0  
85         7.20      0  
100       15.00      1  
109       15.25      1  
117       14.74      1  
124       11.97      0  
131       12.80      0  
136       13.44      0  
144       14.25      1  
148       12.42      0

## IV. Groupbys allow you to segment and aggregate data.

In [44]:
# Segmenting and Aggregating Data with Groupby function

# group dataset by specie

specie_group = iris_df.groupby('species')

#find mean of the new dataset
specie_group.mean()




sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa             5.006        3.428         1.462        0.246      0.3656   
versicolor         5.936        2.770         4.260        1.326      5.7204   
virginica          6.588        2.974         5.552        2.026     11.2962   

            giant  
species            
setosa       0.00  
versicolor   0.00  
virginica    0.12

In [45]:
# display median measurement for each specie

specie_group.median()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa               5.0          3.4          1.50          0.2       0.300   
versicolor           5.9          2.8          4.35          1.3       5.615   
virginica            6.5          3.0          5.55          2.0      11.445   

            giant  
species            
setosa        0.0  
versicolor    0.0  
virginica     0.0

In [46]:
# display standard deviation for all specie

specie_group.std()


sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa          0.352490     0.379064      0.173664     0.105386    0.181155   
versicolor      0.516171     0.313798      0.469911     0.197753    1.368403   
virginica       0.635880     0.322497      0.551895     0.274650    2.157412   

               giant  
species               
setosa      0.000000  
versicolor  0.000000  
virginica   0.328261

In [47]:
# display maximum for each specie

specie_group.max()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa               5.8          4.4           1.9          0.6        0.96   
versicolor           7.0          3.4           5.1          1.8        8.64   
virginica            7.9          3.8           6.9          2.5       15.87   

            giant  
species            
setosa          0  
versicolor      0  
virginica       1

In [48]:
# display minimum for each specie

specie_group.min()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa               4.3          2.3           1.0          0.1        0.11   
versicolor           4.9          2.0           3.0          1.0        3.30   
virginica            4.9          2.2           4.5          1.4        7.50   

            giant  
species            
setosa          0  
versicolor      0  
virginica       0

In [50]:
# display min(), max(), median() mean at the same time using agg()

specie_group.agg(['min', 'max', 'median', 'mean'])

sepal_length                    sepal_width                     \
                    min  max median   mean         min  max median   mean   
species                                                                     
setosa              4.3  5.8    5.0  5.006         2.3  4.4    3.4  3.428   
versicolor          4.9  7.0    5.9  5.936         2.0  3.4    2.8  2.770   
virginica           4.9  7.9    6.5  6.588         2.2  3.8    3.0  2.974   

           petal_length       ... petal_width        petal_area         \
                    min  max  ...      median   mean        min    max   
species                       ...                                        
setosa              1.0  1.9  ...         0.2  0.246       0.11   0.96   
versicolor          3.0  5.1  ...         1.3  1.326       3.30   8.64   
virginica           4.5  6.9  ...         2.0  2.026       7.50  15.87   

                            giant                   
            median     mean   min max median  mean  
species                                             
setosa       0.300   0.3656     0   0    0.0  0.00  
versicolor   5.615   5.7204     0   0    0.0  0.00  
virginica   11.445  11.2962     0   1    0.0  0.12  

[3 rows x 24 columns]

In [51]:
# display min(), max() at the same time using agg()

specie_group.agg(['min', 'max'])

sepal_length      sepal_width      petal_length      petal_width  \
                    min  max         min  max          min  max         min   
species                                                                       
setosa              4.3  5.8         2.3  4.4          1.0  1.9         0.1   
versicolor          4.9  7.0         2.0  3.4          3.0  5.1         1.0   
virginica           4.9  7.9         2.2  3.8          4.5  6.9         1.4   

                petal_area        giant      
            max        min    max   min max  
species                                      
setosa      0.6       0.11   0.96     0   0  
versicolor  1.8       3.30   8.64     0   0  
virginica   2.5       7.50  15.87     0   1

#### <span style="color:#555">EXERCISES</span>

Complete each of the following exercises.

First, remember to import Pandas and read in the iris dataset.

In [1]:
import pandas as pd

df = pd.read_csv('project_files/iris.csv')

## <span style="color:RoyalBlue">Exercise 5.1 - Flowers for Casey</span>

In the previous lesson, we discovered that the first local business we need to visit in Park Royal is a local **flower shop.** When we arrive, we find out that the owner, Casey, urgently needs our help!

They just received a new shipment of **iris flowers**, but they've never stocked these flowers before. Casey asks us to share what we know about these new flowers. Luckily, we have the *iris dataset* to learn more about them ourselves!

In the previous lesson, we also introduced the idea of using a **toy problem** to reduce the size of a problem and break down the concepts. While 150 observations is not exactly "big data," it's still too large to fit on our screen at once. So let's use another toy problem to practice the concepts in this lesson.

#### A.) First, create a new DataFrame called <code style="color:steelblue">toy_df</code>. It should contain the first 5 rows plus the last 5 rows from our original Iris dataset.
* **Tip:** You already have a <code style="color:steelblue">.head()</code>, but what about a <code style="color:steelblue">.tail()</code>?
* **Tip:** <code style="color:steelblue">pd.concat()</code> is your [friend](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html).

In [2]:
# Create toy_df

toy_df = pd.concat([df.head(), df.tail()])


#### B.) Next, display <code style="color:steelblue">toy_df</code>.
* After all, it will only be 10 rows.
* In <code style="color:steelblue">toy_df</code>, you should have data from 2 different species of flower. Which are they?

In [3]:
# Display toy_df
toy_df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

<strong style="color:RoyalBlue">Expected output:</strong>

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
    </tr>
  </tbody>
</table>

#### C.) Next, display a summary table for <code style="color:steelblue">toy_df</code>.
* It should have the mean, standard deviation, and quartiles for each of the columns.
* Since <code style="color:steelblue">toy_df</code> is only 10 rows, you can manually check the **mins** and **maxes**. Are they correct?

In [4]:
# Describe toy_df

toy_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count     10.000000    10.000000     10.000000    10.000000
mean       5.590000     3.130000      3.290000     1.130000
std        0.807534     0.316403      1.995244     0.992248
min        4.600000     2.500000      1.300000     0.200000
25%        4.925000     3.000000      1.400000     0.200000
50%        5.500000     3.050000      3.250000     1.000000
75%        6.275000     3.350000      5.175000     1.975000
max        6.700000     3.600000      5.400000     2.300000

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>5.590000</td>
      <td>3.130000</td>
      <td>3.290000</td>
      <td>1.130000</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.807534</td>
      <td>0.316403</td>
      <td>1.995244</td>
      <td>0.992248</td>
    </tr>
    <tr>
      <th>min</th>
      <td>4.600000</td>
      <td>2.500000</td>
      <td>1.300000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>4.925000</td>
      <td>3.000000</td>
      <td>1.400000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>5.500000</td>
      <td>3.050000</td>
      <td>3.250000</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>6.275000</td>
      <td>3.350000</td>
      <td>5.175000</td>
      <td>1.975000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>6.700000</td>
      <td>3.600000</td>
      <td>5.400000</td>
      <td>2.300000</td>
    </tr>
  </tbody>
</table>

## <span style="color:RoyalBlue">Exercise 5.2 - Basic Feature Engineering</span>

Elementwise operations are very useful in machine learning, especially for **feature engineering.** Feature engineering is the process of creating new features (model input variables) from existing ones, and its one of the best ways for data scientists to add value.

In the Iris dataset, we have petal width and length, but what if we wanted to know petal area? Well, we can create a new <code style="color:steelblue">petal_area</code> feature (yes, the petals are not perfect rectangles, but that's fine).

#### A.) First, display the two columns of <code style="color:steelblue">petal_width</code> and <code style="color:steelblue">petal_length</code> in <code style="color:steelblue">toy_df</code>.
* **Tip:** You can index a DataFrame using a list of column names too, like so:


<pre style="color:steelblue">df[['column_1', 'column_2']]</pre>

In [5]:
# Display petal_width and petal_length


toy_df[['petal_width', 'petal_length']]

petal_width  petal_length
0            0.2           1.4
1            0.2           1.4
2            0.2           1.3
3            0.2           1.5
4            0.2           1.4
145          2.3           5.2
146          1.9           5.0
147          2.0           5.2
148          2.3           5.4
149          1.8           5.1

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>petal_width</th>
      <th>petal_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.2</td>
      <td>1.3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.2</td>
      <td>1.5</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>145</th>
      <td>2.3</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>146</th>
      <td>1.9</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>147</th>
      <td>2.0</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>148</th>
      <td>2.3</td>
      <td>5.4</td>
    </tr>
    <tr>
      <th>149</th>
      <td>1.8</td>
      <td>5.1</td>
    </tr>
  </tbody>
</table>

#### B.) Next, create a new <code style="color:steelblue">petal_area</code> feature in <code style="color:steelblue">toy_df</code>.
* Multiply the <code style="color:steelblue">petal_width</code> column by the <code style="color:steelblue">petal_length</code> column.
* Display <code style="color:steelblue">toy_df</code> after creating the new feature.
* Are the values for <code style="color:steelblue">petal_area</code> correct? Manually spot check a few of them just to make sure.

In [7]:
# Create a new petal_area column

toy_df['petal_area'] = toy_df['petal_width'] * toy_df['petal_length']

# Display toy_df
toy_df


sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
  </tbody>
</table>

### C.) Finally, what do we now know about Iris flowers?

By creating a <code style="color:steelblue">petal_area</code> feature, it's now much easier to see that virginica flowers have significantly larger petals than setosa flowers do! Often, by creating new features, you can learn more about the data (and improve your machine learning models). As you might guess, a machine learning algorithm and take these features and tease out the underlying patterns in a much more precise and comprehensive way.

## <span style="color:RoyalBlue">Exercise 5.3 - Mask On Mask Off</span>

Now, let's use the <code style="color:steelblue">toy_df</code> to really drive home the concept of **boolean masks**, because we'll be using these very frequently in data science.

For example, assume we wanted to display observations where <code style="color:steelblue">petal_area > 10</code> and <code style="color:steelblue">sepal_width > 3</code>. How could we do so?

#### A.) First, display <code style="color:steelblue">toy_df</code> again just to have it in front of you.

In [8]:
# Display toy_df

toy_df


sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18

#### B.) Take a look at the DataFrame and manually count the number that satisfy our conditions.
* How many observations have <code style="color:steelblue">petal_area > 10</code>?
* How many observations have <code style="color:steelblue">sepal_width > 3</code>?
* How many satisfy both conditions?

*In the toy_df, there are:*
* **3** observations with <code style="color:steelblue">petal_area > 10</code> (indices 145, 147, 148)
* **5** observations with <code style="color:steelblue">sepal_width > 3</code> (indices 0,2,3,4,148)
* **1** observation that satifies both conditions (index 148)

Great. Now we'll see what's going on under the hood when we use our boolean masks.

#### C.) Create a boolean mask for <code style="color:steelblue">petal_area > 10</code>.
* Name it <code style="color:steelblue">petal_area_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [9]:
# Mask for petal_area > 10
petal_area_mask = toy_df['petal_area'] > 10

# Display petal_area_mask
petal_area_mask

0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool
</pre>

#### D.) Next, create a boolean mask for <code style="color:steelblue">sepal_width > 3</code>.
* Name it <code style="color:steelblue">sepal_width_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [10]:
# Mask for sepal_width > 3
sepal_width_mask = toy_df['sepal_width'] > 3


# Display sepal_width_mask
sepal_width_mask

0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool
</pre>

#### E.) Next, display the two masks combined using the <code style="color:steelblue">&</code> operator.
* Note how their combination results in another boolean mask!

In [12]:
# Display both masks, combined
petal_area_mask & sepal_width_mask

0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool
</pre>

#### F.) Finally, select the observation(s) from <code style="color:steelblue">toy_df</code> where both conditions are met.
* Do you get the result you expected based on (B)?

In [13]:
# Index with both masks

toy_df[petal_area_mask & sepal_width_mask]


sepal_length  sepal_width  petal_length  petal_width    species  \
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
148       12.42

## <span style="color:RoyalBlue">Exercise 5.4 - Iris Insights</span>

With the help of **groupbys**, we're almost ready to return to Casey and share what we've learned about Iris flowers. But before we do, let's just bring back our <code style="color:steelblue">toy_df</code> for one last hoorah, just to make sure we know what's going on under the hood.

Let's calculate the median <code style="color:steelblue">petal_area</code> for each species. Since <code style="color:steelblue">toy_df</code> is small, we can do this manually as well and check to make sure the values are correct.


#### A.) First, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the virginica flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the virginica species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* **Tip:** Check out the <code style="color:steelblue">.sort_values()</code> function ([documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)).

In [54]:
# Display all 'virginica' species, sorted by petal_area

virginica_mask = toy_df['species'] == 'virginica'

toy_petal_area = toy_df[virginica_mask].sort_values('petal_area')

toy_median = toy_df[virginica_mask]['petal_area'].median()

toy_petal_area
#toy_median

sepal_length  sepal_width  petal_length  petal_width    species  \
149           5.9          3.0           5.1          1.8  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
145           6.7          3.0           5.2          2.3  virginica   
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
149        9.18  
146        9.50  
147       10.40  
145       11.96  
148       12.42

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
  </tbody>
</table>

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the virginica species?

#### B.) Next, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the setosa flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the setosa species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* Based on the output, what's median <code style="color:steelblue">petal_area</code> for the setosa species?

In [41]:
# Display all 'setosa' species
setosa_mask = toy_df['species'] == 'setosa'

toy_df[setosa_mask].sort_values('petal_area')

sepal_length  sepal_width  petal_length  petal_width species  petal_area
2           4.7          3.2           1.3          0.2  setosa        0.26
0           5.1          3.5           1.4          0.2  setosa        0.28
1           4.9          3.0           1.4          0.2  setosa        0.28
4           5.0          3.6           1.4          0.2  setosa        0.28
3           4.6          3.1           1.5          0.2  setosa        0.30

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
  </tbody>
</table>

#### C.) Finally let's calculate the median values using a <code style="color:steelblue">.groupby()</code>.
* Do you get the same result?

In [51]:
# Median petal_area in toy_df

toy_df.groupby('species')['petal_area'].median()

species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64
</pre>

#### Congratulations... You've completed the training mission!

In this lesson, you explored the Iris dataset using Pandas. After arming yourself with Pandas, you've successfully completed the rest of the training mission!*